### PR TITLE
core: Implement constant folding for AffineExpr

### PR DESCRIPTION
**Pull request by Groverkss**

### CODE REVIEW

File: tests/filecheck/parser-printer/affine_map.mlir [hunk 1]
```diff
@@ -27,4 +27,25 @@ builtin.module {
 
   // CHECK: "f9"() {"map" = #affine_map<(d0, d1, d2)[s0, s1, s2] -> ((d0 + s0), (d1 + s1), (d2 + s2))>} : () -> ()
   "f9"() {map = affine_map<(d0, d1, d2)[s0, s1, s2] -> (d0 + s0, d1 + s1, d2 + s2)>} : () -> ()
+
+  // CHECK: "f10"() {"map" = #affine_map<(d0, d1) -> (((d1 + (d0 * -1)) + ((((((d0 + (d1 * -1)) + 1) * 2) + d1) + -1) + 0)), (((((d1 + d1) + 1) + d1) + d1) + 1))>} : () -> ()
+  "f10"() { map = affine_map<(i, j) -> ((j - i) + 2*(i - j + 1) + j - 1 + 0, j + j + 1 + j + j + 1)> } : () -> ()
+
+  // CHECK: "f11"() {"map" = #affine_map<(d0, d1) -> ((((d0 + 4) * -1) + 6), d1)>} : () -> ()
+  "f11"() { map = affine_map<(i, j) -> (3+3-2*2+i, j)>} : () -> ()
```
Comment (webmiche) [inline tests/filecheck/parser-printer/affine_map.mlir:35]: This looks wrong to me. Shouldn't the initial thing translate to 2+i? It seems to translate to 2-i, right?

File: xdsl/ir/affine/affine_expr.py [hunk 2]
```diff
@@ -159,9 +159,42 @@ def eval(self, dims: list[int], symbols: list[int]) -> int:
...
         if isinstance(other, int):
             other = AffineExpr.constant(other)
+
+        # Fold constants.
+        if fold := self._try_fold_constant(other, _AffineExprKind.Add):
+            return fold
+
+        # Keep the constant on the right side.
+        if self._impl.kind == _AffineExprKind.Constant:
+            self, other = other, self
```
Comment (math-fehr) [inline xdsl/ir/affine/affine_expr.py:196]: That's so cursed ;) (But that's fine)

**Review by math-fehr — APPROVED** [review]
Nice! Thanks!

File: tests/filecheck/parser-printer/affine_map.mlir (hunk 1 above)
Comment (superlopuh) [reply tests/filecheck/parser-printer/affine_map.mlir:35]: the whole thing is weird the expression in the check seems more complex than the one in the source...

**Review by superlopuh — COMMENTED** [review]
I don't really understand, it seems like the output is more complex than the source. Should we not aim to at least make it as simple as the input, and maybe simpler?

File: tests/filecheck/parser-printer/affine_map.mlir (hunk 1 above)
Comment (Groverkss) [reply tests/filecheck/parser-printer/affine_map.mlir:35]: Oops, this was a parser bug. Fixed now.

File: xdsl/ir/affine/affine_expr.py [hunk 3]
```diff
@@ -159,9 +159,42 @@ def eval(self, dims: list[int], symbols: list[int]) -> int:
 
         raise ValueError("Unreachable")
 
+    def _try_fold_constant(
```
Comment (superlopuh) [inline xdsl/ir/affine/affine_expr.py:162]: Before you merge this, why not make the signature of this function 
```suggestion
    def _try_fold_constant(self) -> AffineExpr:
```
Where you return self if you could not merge the constant? Where the implementation folds the constant in the lhs, rhs, and then matches on the kind of the operation to see how they could be combined?

Comment (superlopuh) [reply xdsl/ir/affine/affine_expr.py:162]: Also it seems like swapping around the order of arguments might be confusing, feels like that's a separate thing from constant folding to me.

**Review by superlopuh — CHANGES_REQUESTED** [review]

File: xdsl/ir/affine/affine_expr.py (hunk 2 above)
Comment (Groverkss) [reply xdsl/ir/affine/affine_expr.py:208]: Any better way to swap things?

**Comment by Groverkss** [comment]
> I don't really understand, it seems like the output is more complex than the source. Should we not aim to at least make it as simple as the input, and maybe simpler?

This will be done as more simplifications are added. This patch just adds some of them.

File: xdsl/ir/affine/affine_expr.py (hunk 3 above)
Comment (Groverkss) [reply xdsl/ir/affine/affine_expr.py:162]: > Also it seems like swapping around the order of arguments might be confusing, feels like that's a separate thing from constant folding to me.

It's canonicalizing the input for simplifications. Seems pretty standard to me?

Comment (Groverkss) [reply xdsl/ir/affine/affine_expr.py:162]: > Before you merge this, why not make the signature of this function
> 
> Where you return self if you could not merge the constant? Where the implementation folds the constant in the lhs, rhs, and then matches on the kind of the operation to see how they could be combined?

I like the current implementation. It allows me to fold things like (expr + constant) + constant too.

**Comment by Groverkss** [comment]
I added some more simplifications for now, but I don't think I'll add any more simplifications for now.

**Review by superlopuh — APPROVED** [review]

File: xdsl/ir/affine/affine_expr.py (hunk 2 above)
Comment (math-fehr) [reply xdsl/ir/affine/affine_expr.py:208]: No that's the best!
The only curse thing in my opinion is swapping `self`, but that's fine!

Comment (webmiche) [reply xdsl/ir/affine/affine_expr.py:208]: Maybe something like `other.__add__(self)`?

**Pull request merged**